lib/src/workspace.rs

```diff
@@ -366,20 +381,37 @@ impl Workspace {
             )
             .context(&repo_file_path)?;
 
+        // Load the new repo from the file we just wrote.
+        // This gives e.g. GitBackend has an opportunity to check if we
+        // are colocated, given we have a new workspace root.
+        let repo_loader = RepoLoader::init_from_file_system(
+            user_settings,
+            repo_path,
+            store_factories,
+            Some(workspace_root),
+        )
+        .map_err(WorkspaceInitExistingError::LoadStore)?;
+
+        // Now load a repo at the current operation. We go through
+        // OperationId as the Operation struct contains a reference
+        // to the old workspace's Store.
+        let current_op = repo_loader
+            .load_operation(existing_repo.operation().id())
+            .map_err(WorkspaceInitExistingError::LoadOperation)?;
+        let repo = repo_loader
+            .load_at(&current_op)
+            .map_err(WorkspaceInitExistingError::LoadRepo)?;
+
```

Comment:
I meant this could be worked around by the caller by recreating the `WorkspaceCommandHelper` after minimal workspace setup has been done. Can you check if that's doable?

Reply:
This code is the minimal workspace setup being done, and the caller goes on to recreate the WorkspaceCommandHelper using the Workspace & repo returned from this function.

You need to reload the store *before* you call `init_working_copy` so that the store returned in the repo from this function is the same as the rest of the places, and then the changes to the repo can be written out (i.e. the tx.finish at the end of workspace/add.rs can call `git::reset_head()`). If you instead ignore the return values of this function and reload the workspace from disk, the `independent_heads` tests fails.

And you can't do a full WorkspaceCommandHelper load from scratch if you haven't called `init_working_copy`, because in that case there is no working copy to load. So this is the only way.

Reply:
I meant something like
```rust
let (new_workspace, ..) = Workspace::init_workspace_with_existing_repo(..);
let loader_factory = ..; // somehow get workspace loader factory
let new_workspace = loader_factory.create(new_workspace.workspace_root())?.load(..);
```

But yeah, it looks uglier than I thought. It's better to add a bit of mess to the library code.